### PR TITLE
SAMZA-2407: Add job.config.loader.* configs in JobConfig & corresponding env variables in ShellCommandConfig

### PR DESCRIPTION
**Pull request by kw2542**
Design:
https://cwiki.apache.org/confluence/display/SAMZA/SEP-23%3A+Simplify+Job+Runner

Changes:

1. Add job.config.loader.factory to specify ConfigLoaderFactory.
2. Add job.config.loader.properties prefix to specify properties needed by ConfigLoader to fetch config.
3. Add ENV_SUBMISSION_CONFIG enviroment variable to pass job submission configs to Yarn.

API Changes:

N/A. This is part of a series PRs, detailed information will be provided in the last/main PR.
Upgrade Instructions:
N/A. This is part of a series PRs, detailed information will be provided in the last/main PR.
Usage Instructions:
N/A. This is part of a series PRs, detailed information will be provided in the last/main PR.

Tests:
N/A.

### CODE REVIEW

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java [hunk 1]
```diff
@@ -47,6 +47,15 @@
   public static final String CONFIG_REWRITERS = "job.config.rewriters";
   public static final String CONFIG_REWRITER_CLASS = "job.config.rewriter.%s.class";
 
+  /**
+   * job.config.loader.factory specifies {@link ConfigLoaderFactory} to get {@link ConfigLoader}
+   */
+  public static final String CONFIG_LOADER_FACTORY = "job.config.loader.factory";
+  /**
+   * job.config.loader.properties prefix wraps properties needed for {@link ConfigLoader} to loader config.
```
Comment (cameronlee314) [inline samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Do you think it would be helpful to put a small example here about how the prefix will be used?

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java [hunk 2]
```diff
@@ -125,6 +134,16 @@
...
+
+  // Auto-sizing related configs tthat ake precedence over respective sizing confings job.container.count, etc,
+  // *only* when job.autosizing.enabled is true. Otherwise current behavior is maintained.
+  public static final String JOB_AUTOSIZING_ENABLED = "job.autosizing.enabled";
+  public static final String JOB_AUTOSIZING_CONTAINER_COUNT = "job.autosizing.container.count";
+  public static final String JOB_AUTOSIZING_CONTAINER_THREAD_POOL_SIZE = "job.autosizing.container.thread.pool.size";
+  public static final String JOB_AUTOSIZING_CONTAINER_MAX_HEAP_MB = "job.autosizing.container.maxheap.mb";
+  public static final String JOB_AUTOSIZING_CONTAINER_MEMORY_MB = "job.autosizing.container.memory.mb";
+  public static final String JOB_AUTOSIZING_CONTAINER_MAX_CORES = "job.autosizing.container.cpu.cores";
+
```
Comment (cameronlee314) [inline samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: I think @rmatharu was working on merging some of this code in separately. You may want to double check with him on that. We should avoid merging the autosizing code with this PR, just to keep the changes isolated.

File: samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala [hunk 3]
```diff
@@ -25,6 +25,11 @@ object ShellCommandConfig {
    */
   val ENV_COORDINATOR_SYSTEM_CONFIG = "SAMZA_COORDINATOR_SYSTEM_CONFIG"
 
+  /**
+   * This enviorment variable is used to pass configs provided during job submission.
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: 1. s/enviorment/environment
2. Is this a JSON serialized map like `ENV_COORDINATOR_SYSTEM_CONFIG`? If so, could you please specify that in the javadoc?

Comment (bkonold) [reply samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: / whatever the format of the config is, it should be indicated in the javadoc

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java (hunk 2 above)
Comment (kw2542) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Good point, this change is pulled in accidentally.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java (hunk 1 above)
Comment (kw2542) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Added

File: samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala (hunk 3 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: Updated the doc to reflect that this is a JSON serialized map.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java [hunk 4]
```diff
@@ -195,20 +205,20 @@ public int getMonitorPartitionChangeFrequency() {
...
+        if (existingPatternForSystem == null) {
+          newPatternForSystem = Pattern.compile(rewriterRegex.get());
+        } else {
+          newPatternForSystem =
+              Pattern.compile(String.join("|", existingPatternForSystem.pattern(), rewriterRegex.get()));
         }
-      });
+        inputRegexesToMonitor.put(rewriterSystem.get(), newPatternForSystem);
+      }
+    });
```
Comment (cameronlee314) [inline samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Is this an indentation change only? If so, can you please undo it so your diff is just your changes?
Your name would go into the commit history for changing this piece of code, so it would be good to avoid that given that your change is unrelated to cleaning up indentation.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java [hunk 5]
```diff
@@ -348,4 +358,11 @@ public String getCoordinatorStreamFactory() {
     return get(COORDINATOR_STREAM_FACTORY, DEFAULT_COORDINATOR_STREAM_CONFIG_FACTORY);
   }
 
+  /**
+   * Get config loader factory according to the configs
+   * @return full qualified name of {@link ConfigLoaderFactory}
+   */
+  public String getConfigLoaderFactory() {
+    return get(CONFIG_LOADER_FACTORY);
```
Comment (cameronlee314) [inline samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Is this an optional config, or is it required? If it is optional, can you please have it return `Optional<String>`, or add a default value? If required, can you please throw a `ConfigException`?

File: samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala [hunk 6]
```diff
@@ -25,6 +25,11 @@ object ShellCommandConfig {
    */
   val ENV_COORDINATOR_SYSTEM_CONFIG = "SAMZA_COORDINATOR_SYSTEM_CONFIG"
 
+  /**
+   * This enviorment variable is used to pass a JSON serialized map of configs provided during job submission.
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: "environment" is spelled incorrectly

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java (hunk 4 above)
Comment (kw2542) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Reverted. It was automatically corrected by IDEA.

File: samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala (hunk 6 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: Fixed. Nice catch.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java (hunk 5 above)
Comment (kw2542) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: It will eventually be required but optional before we fully deprecate the old path where we read config from coordinator stream. 

I will use Optional<String> for now, and will change to required with ConfigException when we clean up the old flow.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java [hunk 7]
```diff
@@ -47,6 +47,17 @@
...
+  /**
+   * job.config.loader.factory specifies {@link ConfigLoaderFactory} to get {@link ConfigLoader}
+   */
+  public static final String CONFIG_LOADER_FACTORY = "job.config.loader.factory";
+  /**
+   * job.config.loader.properties prefix wraps properties needed for {@link ConfigLoader} to loader config.
+   * e.g. {@link org.apache.samza.config.loaders.PropertiesConfigLoaderFactory}
+   * will read job.config.loader.properties.path to locate the config file location
+   */
+  public static final String CONFIG_LOADER_PROPERTIES_PREFIX = "job.config.loader.properties.";
```
Comment (sborya) [inline samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: where is it used? if it used only in the Loader, should't it be defined there too?

File: samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala [hunk 8]
```diff
@@ -25,6 +25,11 @@ object ShellCommandConfig {
    */
   val ENV_COORDINATOR_SYSTEM_CONFIG = "SAMZA_COORDINATOR_SYSTEM_CONFIG"
 
+  /**
+   * This environment variable is used to pass a JSON serialized map of configs provided during job submission.
+   */
+  val ENV_SUBMISSION_CONFIG = "SAMZA_SUBMISSION_CONFIG"
```
Comment (sborya) [inline samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: where is it used?

Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: This will be used in YarnJob when submitting the job to Yarn. This environment variable will encapsulate config overrides users would like to override on AM.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java (hunk 7 above)
Comment (kw2542) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Yes, it is supposed to be only used by config loader, which specifies properties it needed to retrieve full job config. 

Are you suggesting we move this to ConfigLoader instead?

Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: yea, because if it is not used here - there is no point keeping it here.

File: samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala (hunk 8 above)
Comment (sborya) [reply samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: So code using it will be added later? then it is fine. In general it is easier to read PR if all relevant code is in. But this is fine too.

Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/config/ShellCommandConfig.scala]: Good point, yes, it will be added later. I am splitting them to make PRs relatively small.

File: samza-core/src/main/java/org/apache/samza/config/JobConfig.java (hunk 7 above)
Comment (kw2542) [reply samza-core/src/main/java/org/apache/samza/config/JobConfig.java]: Moved to ConfigLoaderFactory
